File: go/rotate_array.go
package leetcode

// Time complexity: O(n) where n is the length of nums
// Space complexity: O(1)
func rotate(nums []int, k int) {
	k = k % len(nums)
	if k == 0 {
		return
	}

	for i := 0; i < len(nums)/2; i++ {
		nums[i], nums[len(nums)-1-i] = nums[len(nums)-1-i], nums[i]
	}

	for i := 0; i < k/2; i++ {
		nums[i], nums[k-1-i] = nums[k-1-i], nums[i]
	}
	for i := 0; i < (len(nums)-k)/2; i++ {
		nums[i+k], nums[len(nums)-1-i] = nums[len(nums)-1-i], nums[i+k]
	}
}

// Time complexity: O(n) where n is the length of nums
// Space complexity: O(n)
func rotate2(nums []int, k int) {
	k = k % len(nums)
	if k == 0 {
		return
	}

	tail := make([]int, 0, k)
	for i := len(nums) - k; i < len(nums); i++ {
		tail = append(tail, nums[i])
	}

	head := make([]int, 0, len(nums)-k)
	for i := 0; i < len(nums)-k; i++ {
		head = append(head, nums[i])
	}

	for i, el := range tail {
		nums[i] = el
	}
	for i, el := range head {
		nums[i+k] = el
	}
}

File: python/rotate-array.py
from typing import List


class Solution:
    # Time complexity: O(n) where n is the length of nums
    # Space complexity: O(1)
    def rotate(self, nums: List[int], k: int) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        k = k % len(nums)
        if k == 0: return
        
        nums.reverse()
        for i in range(k // 2):
            nums[i], nums[k-1-i] = nums[k-1-i], nums[i]
        for i in range((len(nums)-k) // 2):
            nums[i+k], nums[len(nums)-1-i] = nums[len(nums)-1-i], nums[i+k]
            
        
    # Relatively naive solution
    # Time complexity: O(n)
    # Space complexity: O(n)
    def rotate2(self, nums: List[int], k: int) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        k = k % len(nums)
        if k == 0: return
        tail = nums[len(nums)-k:]
        head = nums[:len(nums)-k]
        for i, el in enumerate(head):
            nums[i+k] = el
        for i, el in enumerate(tail):
            nums[i] = el            
        